ocaml/xapi/cert_distrib.ml

```diff
@@ -364,15 +364,32 @@ let collect_pool_certs ~__context ~rpc ~session_id ~map ~from_hosts =
          map cert
      )
 
-let take_and_append n x xs =
-  (* take_and_append 3 10 [1;2;3;4] = [1;2;3;10] *)
-  let rec loop i acc = function
-    | x :: xs when i < n ->
-        loop (i + 1) (x :: acc) xs
-    | _ ->
-        x :: acc |> List.rev
+(* insert an element randomly *between* existing elements of a list.
+   We are enumerating the gaps between elements in a list:
+
+   List:     a b c d
+   Gaps:    0 1 2 3 4
+
+   A list of length n has gaps 0, 1, ...,n. We are selecting a gap
+   between 1 and n-1 to insert an element - so never at the beginning or
+   end if n >= 2. Otherwise we insert at the front.
+
+   This function is not tail recursive, so only apply to short lists *)
+
+let randomly_insert x ys =
+  let gap = Random.int (List.length ys) + 1 in
```

Comment:
I'm weary of puttin the random pick far away from the seed initialization as it was before

This would allow to call an insert function with clearly defined semantic which should be clearer to understand:
```ocaml
let insert ~at x xs =
  let front, back = Xapi_stdext_std.Listext.List.chop at xs in
  List.concat [front; x :: back]
``` 

this can be used with
```ocaml
let at = match List.length with 
  | 0 -> 0
  | len ->
(* disallow fistpoints at the end of the operation list*)
Random.int (len - 1)
in
let ops' = insert ~at throw_op ops
```

Reply:
I'm fine with this but again believe it is difficult to see what happens around edge cases and this is where the previous problem crept in.

Reply:
The previous code didn't try to handle edge-cases at all. I don't think mixing selection and insertion code is the solution rather, but rather tackling the edge-cases at the selection